kernel/objects/capability-utils/objects/ops.cc

```diff
@@ -41,26 +41,16 @@ namespace mythos {
     auto parentRange = parent.getPtr()->addressRange(parentEntry, parent);
     auto otherRange = other.getPtr()->addressRange(otherEntry, other);
     //MLOG_DETAIL(mlog::cap, DVAR(parentRange), DVAR(otherRange));
-    if (parentRange.contains(otherRange)) {
+    if (!parentRange.contains(otherRange)) return false; // is not within parent's range
+    if (!otherRange.contains(parentRange)) return !parent.isReference(); // subrange is child unless parent is reference
 
-      if (!otherRange.contains(parentRange)) {
-        // parent range is real superset
-        // so it's a child if parent is not a reference
-        return !parent.isReference();
-      }
+    // both cover exactly the same range
+    if (parent.isReference()) return false; // references can only have siblings, this also covers parent.isDerivedReverence()
+    if (parent.isDerived()) return other.isDerivedReference(); // derived can have only derived references as child
```

Comment:
This looks stupid. Consider we first create a child B of same range as original from A and then a derived C from A. Should B be a child of C? This ambiguity can be resolved by prohibiting derived to have originals as children. Which we did here :)

Reply:
Maybe we should document the expected results for the same range with a table in the comments (and maybe the docs).
This should be the right one (and is consistent with this implementation).

| parent | original | reference | derived | derived ref |
|----|---|---|---|---|
| original | x | x | x | panic |
| reference | - | - | - | - |
| derived | - | - | - | x |
| derived ref | - | - | - | - |